File: cheb/cheb.py
"""
Python wrapper functions for taking Chebyshev pseudospectral derivatives
over the interval [lower, upper].
"""
import os, ctypes, pathlib, glob
import numpy as np

# path of the shared library
_path_cheb = glob.glob(str(pathlib.Path(__file__).parent.parent)+'/build/*/*/cheb.so')[0] 
_lib_cheb = ctypes.CDLL(str(_path_cheb))

_cheb_initialized = False
#=============================================================================
_cheb_init          = _lib_cheb.init 
_cheb_init.restype  = None
_cheb_init.argtypes = [
      ctypes.c_size_t,
      ctypes.c_double, 
      ctypes.c_double
      ]

def init(n:int, lower:float, upper:float) -> None:
   """
   Initialize internal FFTW pointers, etc.
   """
   global _cheb_initialized
   _cheb_initialized = True
   _cheb_init(
         ctypes.c_size_t(n), 
         ctypes.c_double(lower), 
         ctypes.c_double(upper)
         )
#=============================================================================
_cheb_cleanup          = _lib_cheb.cleanup
_cheb_cleanup.restype  = None
_cheb_cleanup.argtypes = []

def cleanup() -> None:
   """
   Free internal FFTW pointers, etc.
   """
   global _cheb_initialized
   assert(_cheb_initialized)
   _cheb_initialized = False
   _cheb_cleanup()
#=============================================================================
_cheb_n          = _lib_cheb.n
_cheb_n.restype  = ctypes.c_size_t
_cheb_n.argtypes = []

def n() -> int:
   """
   Number of Chebyshev collocation points.
   """
   assert(_cheb_initialized)
   return _cheb_n()
#=============================================================================
_cheb_lower          = _lib_cheb.lower
_cheb_lower.restype  = ctypes.c_double
_cheb_lower.argtypes = []

def lower() -> float:
   """
   Lower boundary of domain in real space.
   """
   assert(_cheb_initialized)
   return _cheb_lower()
#=============================================================================
_cheb_upper          = _lib_cheb.upper
_cheb_upper.restype  = ctypes.c_double
_cheb_upper.argtypes = []

def upper() -> float:
   """
   Upper boundary of domain in real space.
   """
   assert(_cheb_initialized)
   return _cheb_upper()
#=============================================================================
_cheb_pt          = _lib_cheb.pt
_cheb_pt.restype  = ctypes.c_double
_cheb_pt.argtypes = [
      ctypes.c_size_t
      ]

def pt(i:int) -> float:
   """
   Location of ith Chebyshev point in real space.
   """
   assert(_cheb_initialized)
   assert(i>=0 and i<n())
   return _cheb_pt(ctypes.c_size_t(i))
#=============================================================================
_cheb_der          = _lib_cheb.der
_cheb_der.restype  = None 
_cheb_der.argtypes = [
      np.ctypeslib.ndpointer(ctypes.c_double),
      np.ctypeslib.ndpointer(ctypes.c_double)
      ]

def der(v:np.array, dv:np.array) -> None:
   """
   Derivative over interval [lower, upper].
   """
   assert(_cheb_initialized)
   assert(v.size==n() and dv.size==n())
   _cheb_der(v,dv)
#=============================================================================
_cheb_filter          = _lib_cheb.filter
_cheb_filter.restype  = None
_cheb_filter.argtypes = [
      np.ctypeslib.ndpointer(ctypes.c_double)
      ]

def filter(v:np.array) -> None:
   """
   Low pass filter in Chebyshev space.
   """
   assert(_cheb_initialized)
   assert(v.size==n())
   _cheb_filter(v)

File: cheb/cheb.c
/*
 * Routines for computing pseudospectral Chebyshev derivatives over
 * interval [lower_, upper_]
 *
 */ 
#include <stddef.h>
#include <math.h>
#include <assert.h>
#include <fftw3.h>
/*
 * static global variables
 */
static size_t n_;

static double lower_;
static double upper_;
static double jacobian_; 
static double *pts_; /* Chebyshev collocation
                        points over interval [lower_,upper_]. */

static double *low_pass_; /* For low pass filter in Chebyshev space. */

static fftw_plan plan_dct_; /* For fftw Fourier transform. */
/*==========================================================================*/
/* setup fftw and vectors */
/*==========================================================================*/
void init(const size_t n, const double lower, const double upper) 
{
   n_ = n;
   lower_ = lower;
   upper_ = upper;
   jacobian_ = (upper_-lower_) / 2.0;

   pts_         = (double *)fftw_malloc(n_*sizeof(double));
   low_pass_    = (double *)fftw_malloc(n_*sizeof(double));
   /* 
    * Chebyshev points on interval [lower, upper]
    */
   for (size_t i=0; i<n_; i++) {
      pts_[i] = jacobian_*cos(M_PI*i/(n_-1.0)) + ((upper_+lower_)/2.0);
   }
   for (size_t i=0; i<n_; i++) {
      low_pass_[i] = exp(-40.0*pow((double)i/n_,10));
   }   
   /* 
    * initialize fftw discrete cosine transform plan 
    */
   double *in_  = (double *)fftw_malloc(n_*sizeof(double));
   double *out_ = (double *)fftw_malloc(n_*sizeof(double));
   assert(in_ !=NULL);
   assert(out_!=NULL);

   for (size_t i=0; i<n_; i++) {
      in_[i]  = 0;
      out_[i] = 0;
   }
   plan_dct_ = fftw_plan_r2r_1d(n_,in_,out_,FFTW_REDFT00,FFTW_PATIENT);
   assert(plan_dct_!=NULL);

   fftw_free(in_);
   fftw_free(out_);
}
/*==========================================================================*/
void cleanup(void)
{
   assert(pts_        !=NULL);
   assert(low_pass_   !=NULL);

   fftw_free(pts_);
   fftw_free(low_pass_);

   assert(plan_dct_!=NULL);
   fftw_destroy_plan(plan_dct_);
   fftw_cleanup();
}
/*==========================================================================*/
/* Functions for returning static variables */
/*==========================================================================*/
size_t n(void) { return n_; }
double lower(void) { return lower_; }
double upper(void) { return upper_; }
double pt(const size_t i) { return pts_[i]; }
/*==========================================================================*/
/* position space to Chebyshev space */
/*==========================================================================*/
void to_ch(double *po, double *ch) 
{
   assert(po!=NULL);
   assert(ch!=NULL);

   fftw_execute_r2r(plan_dct_, po, ch);
   /*
    * normalize Chebyshev coefficients
    */
   ch[0]    /= (2.0*(n_-1));
   ch[n_-1] /= (2.0*(n_-1));
   for (size_t i=1; i<n_-1; i++) {
      ch[i] /= (n_-1.0);
   }
}
/*==========================================================================*/
/* Chebyshev space to position space */
/*==========================================================================*/
void to_po(double *ch, double *po) 
{
   assert(ch!=NULL);
   assert(po!=NULL);

   /*
    * normalize coefficients for Fourier transform 
    */
   for (size_t i=1; i<n_-1; i++) {
      ch[i] /= 2.0;
   }
   fftw_execute_r2r(plan_dct_, ch, po);
}
/*==========================================================================*/
/* Compute derivative over interval */
/*==========================================================================*/
void der(double *v, double *dv)
{
   assert(v!=NULL);
   assert(dv!=NULL);
   double *ch = (double *)fftw_malloc(n_*sizeof(double));
   assert(ch!=NULL);

   to_ch(v,ch);
   /*
    * to start Chebyshev derivative recurrence relation 
    */
   ch[n_-1] = 0;
   ch[n_-2] = 0;
   /* 
    * use dv as a temporary array
    */
   for (size_t i=0; i<n_; i++) {
      dv[i] = ch[i];
   }
   /* 
    * apply Chebyshev derivative recurrence relation 
    */
   for (size_t i=n_-2; i>=1; i--) { 
      ch[i-1] = 2.0*i*dv[i] + ch[i+1];
   } 
   ch[0] /= 2.0;
   /* 
    * Normalize derivative to inverval 
    */
   to_po(ch,dv);
   for (size_t i=0; i<n_; i++) {
      dv[i] /= jacobian_;
   }
   fftw_free(ch);
}
/*==========================================================================*/
/* Low pass filter of Chebyshev coefficients */
/*==========================================================================*/
void filter(double *v)
{
   assert(v!=NULL);
   double *ch = (double *)fftw_malloc(n_*sizeof(double));
   assert(ch!=NULL);

   to_ch(v,ch);
   for (size_t i=0; i<n_; i++) { 
      ch[i] *= low_pass_[i];
   } 
   to_po(ch,v);

   fftw_free(ch);
}
/*==========================================================================*/
